File: programers_Lv1/모의고사.js
// https://school.programmers.co.kr/learn/courses/30/lessons/42840?language=javascript
function solution(answers) {
  const studentsInfo = [
    { pick: [1, 2, 3, 4, 5], pickLength: 5 },
    {
      pick: [2, 1, 2, 3, 2, 4, 2, 5],
      pickLength: 8,
    },
    { pick: [3, 3, 1, 1, 2, 2, 4, 4, 5, 5], pickLength: 10 },
  ];

  const answersLength = answers.length;

  studentsInfo.forEach((v, i) => {
    if (answersLength > v.pickLength) {
      studentsInfo[i].pick = Array.from(
        {
          length:
            answersLength % studentsInfo[i].pickLength === 0
              ? answersLength / studentsInfo[i].pickLength
              : Number(answersLength / studentsInfo[i].pickLength) + 1,
        },
        () => studentsInfo[i].pick
      ).flat();
    }
  });

  const scoreArr = [0, 0, 0];

  const [student1, student2, student3] = studentsInfo;
  const pickArr = [student1.pick, student2.pick, student3.pick];
  answers.forEach((v, i) => {
    pickArr.forEach((w, j) => {
      if (v === w[i]) {
        scoreArr[j]++;
      }
    });
  });

  let maxScore = Math.max(...scoreArr);

  return scoreArr
    .map((el, i) => (el === maxScore ? i + 1 : -1))
    .filter((index) => index !== -1);
}

console.log(solution([1, 3, 2, 4, 2]));

File: programers_Lv1/모의고사.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42840?language=python3
def solution(answers):
    students = [[1,2,3,4,5], [2,1,2,3,2,4,2,5], [3,3,1,1,2,2,4,4,5,5]]

    students_pattern_len = []
    for lst in students:
        students_pattern_len.append(len(lst))
    
    answers_len = len(answers)

    for i, v in enumerate(students_pattern_len):
        if answers_len > v:
            students[i] = students[i] * (answers_len // students_pattern_len[i] if answers_len % students_pattern_len[i] == 0 else answers_len // students_pattern_len[i] + 1)
    
    count_list = [0, 0, 0]
    for i, v in enumerate(answers):
        for j in range(len(count_list)):
            if students[j][i] == v:
                count_list[j] += 1
        
    max_score = max(count_list)
    return list(map(lambda x: x+1, filter(lambda x: count_list[x] == max_score, range(len(count_list)))))

# print(solution([1,2,3,4,5]))
print(solution([1,3,2,4,2]))